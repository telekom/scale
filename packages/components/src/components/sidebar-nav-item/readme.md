# scale-sidebar-nav-item



<!-- Auto Generated Below -->


## Properties

| Property       | Attribute       | Description                                                                                                             | Type                | Default     |
| -------------- | --------------- | ----------------------------------------------------------------------------------------------------------------------- | ------------------- | ----------- |
| `active`       | `active`        | Text gets the active color                                                                                              | `boolean`           | `false`     |
| `bold`         | `bold`          | Bold text                                                                                                               | `boolean`           | `false`     |
| `condensed`    | `condensed`     | Used normally for third level items, remove the bottom border                                                           | `boolean`           | `false`     |
| `current`      | `current`       | Mark the child link as "current" with `aria-current=page`. Provide the text hint if needed, default is: "Zurzeit aktiv" | `boolean \| string` | `null`      |
| `nestingLevel` | `nesting-level` | Nesting level within the <scale-sidebar-nav> parent, gets set automatically                                             | `number`            | `undefined` |
| `styles`       | `styles`        | (optional) Extra styles                                                                                                 | `string`            | `undefined` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
