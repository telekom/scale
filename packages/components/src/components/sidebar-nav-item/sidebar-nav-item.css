/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

:host {
  /* FIXME:TOKEN previously --scl-color-grey-10 */
  --border-bottom-color: var(--telekom-color-ui-subtle);
  /* FIXME:TOKEN previously --scl-color-grey-70 */
  --border-left-color-third-nesting: var(--telekom-color-ui-strong);
  --color-active: var(--telekom-color-text-and-icon-primary-pressed);
  /* FIXME:TOKEN this variable is used sometimes for background soetimes for font */
  --color-primary: var(--telekom-color-text-and-icon-primary-standard);
  --background-color-before-active: var(--color-primary);
  --font-weight-bold: var(--telekom-typography-font-weight-bold);
  --box-shadow-focus: inset 0 0 0 var(--telekom-spacing-unit-x05)
    var(--telekom-color-functional-focus);
}
.sidebar-nav-item {
  border-bottom-style: solid;
  border-bottom-width: 1px;
  border-bottom-color: var(--border-bottom-color);
}

.sidebar-nav-item--active {
  position: relative;
  color: var(--color-primary);
}
.sidebar-nav-item--active::before {
  content: '';
  display: block;
  position: absolute;
  top: 0;
  left: var(--left-current-border, 0);
  width: 0; /* The transparent border makes the 2px */
  height: 100%;
  background-color: var(--background-color-before-active);
  /* Accessibility: extra transparent line for Windows HCM */
  border-left: 2px solid transparent;
}

.sidebar-nav-item--condensed.sidebar-nav-item--active::before {
  left: -1px;
}

:host([nesting-level='1']) .sidebar-nav-item,
.sidebar-nav-item--bold {
  font-weight: var(--font-weight-bold);
}

:host([nesting-level='3']) .sidebar-nav-item {
  border-left-color: var(--border-left-color-third-nesting);
  border-left-style: solid;
  border-left-width: 0.0625rem;
}

::slotted(a) {
  color: currentColor;
  display: block;
  padding-top: 1rem;
  padding-right: 1rem;
  padding-bottom: 1rem;
  padding-left: var(--spacing-indent);
  /* FIXME:TOKEN previously --scl-radius-2 no equivalent */
  border-radius: var(--telekom-radius-small);
  text-decoration: none;
  outline: none;
}

:host([nesting-level='2']) ::slotted(a) {
  padding-left: calc(var(--spacing-indent) + 0.75rem);
}

::slotted(a:hover) {
  color: var(--color-primary);
}

::slotted(a:focus) {
  box-shadow: var(--box-shadow-focus);
}

::slotted(a:active) {
  color: var(--color-active);
}

.sidebar-nav-item--condensed {
  border-bottom-width: 0;
}

.sidebar-nav-item--condensed ::slotted(a) {
  padding-top: 0.5rem;
  padding-left: 0.75rem;
  padding-right: 0.5rem;
  padding-bottom: 0.5rem;
}
