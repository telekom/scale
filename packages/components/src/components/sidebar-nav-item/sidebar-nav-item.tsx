/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { Component, h, Prop, Host, Watch, Element } from '@stencil/core';
import classNames from 'classnames';

const SR_ACTIVE_TEXT = ' Zurzeit aktiv';

const isActive = (current) => {
  try {
    return !!JSON.parse(current);
  } catch (e) {
    if (typeof current === 'string') {
      return true;
    }
    return !!current;
  }
};

const getScreenReaderText = (current) => {
  let text;
  try {
    text = JSON.parse(current);
  } catch (e) {
    text = current;
  }

  return typeof text === 'string' && text.length > 0
    ? ` ${text}`
    : SR_ACTIVE_TEXT;
};

@Component({
  tag: 'scale-sidebar-nav-item',
  styleUrl: 'sidebar-nav-item.css',
  shadow: true,
})
export class SidebarNavItem {
  srOnlyElement: HTMLElement;

  @Element() el: HTMLElement;

  /** Used normally for third level items, remove the bottom border */
  @Prop() condensed: boolean = false;
  /** Bold text */
  @Prop() bold: boolean = false;
  /** Text gets the active color */
  @Prop({ mutable: true, reflect: true }) active: boolean = false;
  /**
   * Mark the child link as "current" with `aria-current=page`.
   * Provide the text hint if needed, default is: "Zurzeit aktiv"
   */
  @Prop() current: string | null | boolean = null;
  /** Nesting level within the <scale-sidebar-nav> parent, gets set automatically */
  @Prop() nestingLevel: number;
  /** (optional) Extra styles */
  @Prop() styles?: string;

  @Watch('nestingLevel')
  nestingLevelChanged(newValue: number) {
    if (newValue === 1) {
      this.bold = true;
    }
  }

  @Watch('current')
  currentChanged(newValue) {
    this.handleAriaCurrentInSlottedA(newValue);
    this.syncActiveToCurrent(newValue);
  }

  componentDidLoad() {
    this.handleAriaCurrentInSlottedA(this.current);
    if (this.current) {
      this.syncActiveToCurrent(this.current);
    }
  }

  /**
   * If an item is `current`, it should be `active` as well
   */
  syncActiveToCurrent(newValue) {
    this.active = isActive(newValue);
  }

  /**
   * When `current` is set, this will:
   * - set the aria-current=page attribute on the link
   * - append a text-only hint for screen readers
   * so we end up with something like this:
   * <a href="..." aria-current="page">
   *    Example<span style="...visible to SR only..."> Active link</span>
   * </a>
   * @param current this.current
   */
  handleAriaCurrentInSlottedA(current) {
    const a = this.el.querySelector('a');

    if (this.srOnlyElement != null) {
      a.removeChild(this.srOnlyElement);
      this.srOnlyElement = null;
    }
    if (a != null) {
      a.removeAttribute('aria-current');
    }
    if (isActive(current) && a != null) {
      this.srOnlyElement = this.createScreenReaderOnlySpan();
      a.appendChild(this.srOnlyElement);
      a.setAttribute('aria-current', 'page');
    }
  }

  createScreenReaderOnlySpan() {
    const text = getScreenReaderText(this.current);
    const span = document.createElement('span');
    // .sr-only but inline
    Object.assign(span.style, {
      position: 'absolute',
      left: '-10000px',
      overflow: 'hidden',
    });
    span.textContent = text;

    return span;
  }

  render() {
    return (
      <Host>
        {this.styles && <style>{this.styles}</style>}
        <li
          part={this.getBasePartMap()}
          class={this.getCssClassMap()}
          role="listitem"
        >
          <slot />
        </li>
      </Host>
    );
  }

  getBasePartMap() {
    return this.getCssOrBasePartMap('basePart');
  }

  getCssClassMap() {
    return this.getCssOrBasePartMap('css');
  }

  getCssOrBasePartMap(mode: 'basePart' | 'css') {
    const component = 'sidebar-nav-item';
    const prefix = mode === 'basePart' ? '' : `${component}--`;

    return classNames(
      component,
      this.bold && `${prefix}bold`,
      this.condensed && `${prefix}condensed`,
      this.active && `${prefix}active`
    );
  }
}
