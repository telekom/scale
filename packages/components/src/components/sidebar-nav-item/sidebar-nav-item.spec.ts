/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { newSpecPage } from '@stencil/core/testing';
import { SidebarNavItem } from './sidebar-nav-item';

describe('SidebarNavItem', () => {
  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [SidebarNavItem],
      html: `<scale-sidebar-nav-item></scale-sidebar-nav-item>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [SidebarNavItem],
      html: `
        <scale-sidebar-nav-item>
          <a href="#">Some foo</a>
        </scale-sidebar-nav-item>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should match snapshot after using currentChanged()', async () => {
    const page = await newSpecPage({
      components: [SidebarNavItem],
      html: `
        <scale-sidebar-nav-item>
          <a href="#">Some foo</a>
        </scale-sidebar-nav-item>`,
    });

    page.rootInstance.currentChanged('anytext');
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
    page.rootInstance.currentChanged(null);
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });

  it('should match snapshot after using createScreenReaderOnlySpan()', async () => {
    const page = await newSpecPage({
      components: [SidebarNavItem],
      html: `
        <scale-sidebar-nav-item>
          <a href="#">Some foo</a>
        </scale-sidebar-nav-item>`,
    });
    page.rootInstance.current = 'currentText';
    page.rootInstance.createScreenReaderOnlySpan();
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
    page.rootInstance.current = '';
    page.rootInstance.createScreenReaderOnlySpan();
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });

  it('check default props', async () => {
    const page = await newSpecPage({
      components: [SidebarNavItem],
      html: `<scale-sidebar-nav-item></scale-sidebar-nav-item>`,
    });
    expect(page.rootInstance.condensed).toBe(false);
    expect(page.rootInstance.bold).toBe(false);
    expect(page.rootInstance.active).toBe(false);
    expect(page.rootInstance.current).toBeNull();
  });

  it('check props being set', async () => {
    const page = await newSpecPage({
      components: [SidebarNavItem],
      html: `<scale-sidebar-nav-item></scale-sidebar-nav-item>`,
    });
    page.root.condensed = true;
    page.root.bold = true;
    page.root.active = true;
    page.root.current = 'current';
    page.root.nestingLevel = 2;
    await page.waitForChanges();
    expect(page.rootInstance.condensed).toBe(true);
    expect(page.rootInstance.bold).toBe(true);
    expect(page.rootInstance.active).toBe(true);
    expect(page.rootInstance.current).toBe('current');
    expect(page.rootInstance.nestingLevel).toBe(2);
  });

  describe('classes', () => {
    it('should handle getCssClassMap() and getBasePartMap()', () => {
      const element = new SidebarNavItem();
      element.bold = true;
      element.condensed = true;
      element.active = true;
      expect(element.getCssClassMap()).toContain('sidebar-nav-item');
      expect(element.getCssClassMap()).toContain('sidebar-nav-item--bold');
      expect(element.getCssClassMap()).toContain('sidebar-nav-item--condensed');
      expect(element.getCssClassMap()).toContain('sidebar-nav-item--active');

      expect(element.getBasePartMap()).toContain('sidebar-nav-item');
      expect(element.getBasePartMap()).toContain('bold');
      expect(element.getBasePartMap()).toContain('condensed');
      expect(element.getBasePartMap()).toContain('active');
    });
  });

  it('should set bold to true if nestingLevel is set to 1', () => {
    const element = new SidebarNavItem();
    expect(element.bold).toBe(false);
    element.nestingLevelChanged(1);
    expect(element.bold).toBe(true);
  });

  it('should set active to true if Watcher currentChanged() is called with a string', () => {
    const element = new SidebarNavItem();
    expect(element.active).toBe(false);
    element.currentChanged('anyString');
    expect(element.active).toBe(true);
  });

  it('should set active to true if componentDidLoad() is called and current has a value', () => {
    const element = new SidebarNavItem();
    element.current = 'anyString';
    element.componentDidLoad();
    expect(element.active).toBe(true);
  });

  /*  it('should set active to true if componentDidLoad() is called and current has a value', () => {
    const element = new SidebarNavItem();
    element.current = 'anyString'
    expect(element.createScreenReaderOnlySpan()).toContainElement()
  }); */
});
