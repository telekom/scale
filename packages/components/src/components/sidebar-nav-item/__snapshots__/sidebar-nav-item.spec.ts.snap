// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarNavItem should match snapshot 1`] = `
<scale-sidebar-nav-item>
  <mock:shadow-root>
    <li class="sidebar-nav-item" part="sidebar-nav-item" role="listitem">
      <slot></slot>
    </li>
  </mock:shadow-root>
</scale-sidebar-nav-item>
`;

exports[`SidebarNavItem should match snapshot 2`] = `
<scale-sidebar-nav-item>
  <mock:shadow-root>
    <li class="sidebar-nav-item" part="sidebar-nav-item" role="listitem">
      <slot></slot>
    </li>
  </mock:shadow-root>
  <a href="#">
    Some foo
  </a>
</scale-sidebar-nav-item>
`;

exports[`SidebarNavItem should match snapshot after using createScreenReaderOnlySpan() 1`] = `
<scale-sidebar-nav-item active="">
  <mock:shadow-root>
    <li class="sidebar-nav-item sidebar-nav-item--active" part="sidebar-nav-item active" role="listitem">
      <slot></slot>
    </li>
  </mock:shadow-root>
  <a aria-current="page" href="#">
    Some foo
    <span style="position: absolute; left: -10000px; overflow: hidden;">
      currentText
    </span>
  </a>
</scale-sidebar-nav-item>
`;

exports[`SidebarNavItem should match snapshot after using createScreenReaderOnlySpan() 2`] = `
<scale-sidebar-nav-item active="">
  <mock:shadow-root>
    <li class="sidebar-nav-item sidebar-nav-item--active" part="sidebar-nav-item active" role="listitem">
      <slot></slot>
    </li>
  </mock:shadow-root>
  <a aria-current="page" href="#">
    Some foo
    <span style="position: absolute; left: -10000px; overflow: hidden;">
      Zurzeit aktiv
    </span>
  </a>
</scale-sidebar-nav-item>
`;

exports[`SidebarNavItem should match snapshot after using currentChanged() 1`] = `
<scale-sidebar-nav-item active="">
  <mock:shadow-root>
    <li class="sidebar-nav-item sidebar-nav-item--active" part="sidebar-nav-item active" role="listitem">
      <slot></slot>
    </li>
  </mock:shadow-root>
  <a aria-current="page" href="#">
    Some foo
    <span style="position: absolute; left: -10000px; overflow: hidden;">
      Zurzeit aktiv
    </span>
  </a>
</scale-sidebar-nav-item>
`;

exports[`SidebarNavItem should match snapshot after using currentChanged() 2`] = `
<scale-sidebar-nav-item>
  <mock:shadow-root>
    <li class="sidebar-nav-item" part="sidebar-nav-item" role="listitem">
      <slot></slot>
    </li>
  </mock:shadow-root>
  <a href="#">
    Some foo
  </a>
</scale-sidebar-nav-item>
`;
